RearrangeCarGroup8Movements.py

```diff
@@ -0,0 +1,195 @@
+        self.car_position_dict = {car:pos for pos, car in enumerate(init_car_order) \
+                                  if car} #search the index of car in O(1) via dict
+        self.not_arranged_cars = {car for i, car in enumerate(init_car_order)\
+                                  if init_car_order[i] != final_car_order[i] and car}
+
+    def _create_log_file(self):
+
+        '''In case the number of moves is more than 1000
+        writes the output to the logfile instead of the console
+        '''
+
+        num_cars = len(self.not_arranged_cars)
+        info_message = 'Number of cars to arrange is {0} > 1000, movements are logged'.format(num_cars)
+        logging_message = info_message
+        logging.warning(logging_message)
+        logging.info('Initial car order:')
+        logging.info(self.initial_car_order)
+        logging.info('Final car order:')
+        logging.info(self.final_car_order)
+        logging.info('Moves:')
+
+    def _move_car_to_empty_slot(self, car_to_move_position):
+
+        '''Arranges the car with car_to_move_position index to the empty slot
+        Arguments:
+            car_to_move_position: the position from where the car is moved
+        Returns:
+            None
+        '''
+
+        car_to_move = self.current_car_order[car_to_move_position]
+        self.current_car_order[self.empty_position] = car_to_move
+        self.current_car_order[car_to_move_position] = 0
+
+        self.car_position_dict[car_to_move] = self.empty_position
+        self.empty_position = car_to_move_position
+
+    def _get_not_arranged_car_index(self):
+
+        '''Finds first car index not yet standing on the final destination
+        '''
+
+        car = random.choice(tuple(self.not_arranged_cars)) # O(1)
+        return self.car_position_dict[car]
+
+    def _rearrange_one_car(self):
+
+        '''Puts to the current empty space the car
+        that should be on this place according to the final arrangement
+        if current empty space position is different from the final order empty
+        space position
+
+        Otherwise picks up the random not arranged car and places it to the empty space
+        in this case the number of not arranged cars stays the same
+
+        Complexity is O(1)
+        '''
+
+        if self.final_car_order[self.empty_position] == 0:
+            car_to_move_position = self._get_not_arranged_car_index()
+            car_to_move = self.current_car_order[car_to_move_position]
+        else:
+            car_to_move = self.final_car_order[self.empty_position]
+            car_to_move_position = self.car_position_dict[car_to_move]
+            self.not_arranged_cars.difference_update([car_to_move]) # O(1)
+
+
+        movement = Movement(car_to_move, car_to_move_position, self.empty_position)
+        if self.print_console:
+            print movement
+        else:
+            movement.print_log_file()
+
+        self._move_car_to_empty_slot(car_to_move_position)

```

Comment:
If we are to return a list of movement objects instead of printing right away, then perhaps we can remove line 151 - 154 and then just return `Movement(car_to_move, car_to_move_position, self.empty_position)` ^^